docs/architecture/adr-075-rpc-subscription.md

```diff
@@ -0,0 +1,668 @@
+}
+
+type Filter struct {
+    Query string `json:"query"`
+}
+```
+
+> **Open question:** For the initial implementation, the server will not cache
+> filter queries for the client. If this turns out to be a problem, we can add
+> methods to save and remove stored queries. If necessary, the Query type can be
+> extended to include a query ID option.
+
+> **Open question:** For the initial implementation, I am planning to keep the
+> existing query language as-is. Another option would be to compile queries on
+> the client side, and provide a more structured representation in the request
+> than just a plain string.
+>
+> I think this is probably worth doing, even if only to offload the parsing
+> step onto the client. For now, though, I've stuck to the existing format to
+> make migration easier. The service can keep a small shared cache of compiled
+> queries if performance turns out to be an issue.
+
+The semantics of the request are as follows: An item in the event log is
+**eligible** for a query if:
+
+- It is newer than the `after_item` cursor (if set).
+- It is older than the `before_item` cursor (if set).
+- It matches the filter (if set).
+
+Among the eligible items in the log, the server returns up to `max_results` of
+the newest items, in reverse order of cursor. If `max_results` is unset the
+server chooses a number to return, and will cap `max_results` at a sensible
+limit.
+
+The `wait_time` parameter is used to effect polling. If `before_item` is empty,
+the server will wait for up to `wait_time` for additional items, if there are
+fewer than `max_results` eligible results in the log.  If `wait_time` is zero,
+the server will return whatever eligible items are available immediately.
+
+If `before_item` non-empty, `wait_time` is ignored: new results are only added
+to the head of the log, so there is no need to wait.  This allows the client to
+poll for new data, and "page" backward through matching event items. This is
+discussed in more detail below.
+
+The server will set a sensible cap on the maximum `wait_time`, overriding
+client-requested intervals longer than that.
+
+A successful reply from the `events` request corresponds to the following Go
+types:
+
+```go
+type EventReply struct {
+    // The items matching the request parameters, from newest
+    // to oldest, if any were available within the timeout.
+    Items []*EventItem `json:"items"`
+
+    // This is true if there is at least one older matching item
+    // available in the log that was not returned.
+    More bool `json:"more"`
+
+    // The cursor of the oldest item in the log at the time of this reply,
+    // or "" if the log is empty.
+    Oldest string `json:"oldest_item"`
+
+    // The cursor of the newest item in the log at the time of this reply,
+    // or "" if the log is empty.
+    Newest string `json:"newest_item"`
+}
+
+type EventItem struct {
+    // The cursor of this item.
+    Cursor string `json:"cursor"`
```

Comment:
what's the format of this (UUDI, or something?)

Reply:
A time-based UUID almost works, but the endian order is wrong to use standard ones. I discussed it a bit [here](https://github.com/tendermint/tendermint/blob/mjf/adr-rpc-sub/docs/architecture/adr-075-rpc-subscription.md#L256), but broadly, a custom format.

See also https://github.com/tendermint/tendermint/blob/mjf/adr-rpc-sub/docs/architecture/adr-075-rpc-subscription.md#L484

Does that answer your question? I can talk more about this in the doc if that helps.

Reply:
What's the lifetime of this cursor? Will this be in memory only? I guess this is closely related to the load balancer question.

Reply:
> What's the lifetime of this cursor? Will this be in memory only? I guess this is closely related to the load balancer question.

A cursor must be unique among all events published by a single node, but it is not required to be unique globally across nodes. It need not be only in-memory (though it will, for the initial implementation of this ADR), but it is not meaningful outside the context of that node.

Morally, a cursor could be a plain sequence number, but it's convenient to use a timestamp simply for implementation purposes. The only properties the client can depend on are that lexicographic order of cursors respects order of publication on that node.

Reply:
In re: Lifetime. In principle, a cursor is valid for its node "forever", but in practice if the node restarts (or crashes, or whatever), it would behave after restart as if the event named by that cursor had expired from the event log.

See also the discussion above about persisting events. The current subscription API does not do this, and this design doesn't propose to add that. (It could be done, but it's not clear it makes sense)